File: C/singlyLLL.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node *next;
};

void insertfirst(struct node ** first,int no)
{
    struct node * newn = NULL;

    newn = (struct node *)malloc(sizeof(struct node));

    newn-> data = no;
    newn-> next = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        newn-> next = *first;
        *first = newn;
    }
}

void insertlast(struct node ** first,int no)
{
    struct node * newn = NULL;
    struct node * temp = NULL;

    newn = (struct node *)malloc (sizeof(struct node));

    newn->data = no;
    newn->next = NULL;

    if(*first == NULL)
    {
        *first = newn;
    }
    else
    {
        temp = *first;

        while(temp->next != NULL)
        {
            temp = temp-> next;
        }
        temp -> next = newn;

    }


}

void deletefirst(struct node ** first)
{
    struct node * temp= NULL;

   if(*first == NULL)
   {
        return ;
   }
   else if( (*first)-> next == NULL)
   {
        free(*first);
        *first = NULL;
   }
   else
   {
       temp = *first;
       *first = temp -> next;
       free(temp);
       
   }
}

void deletelast(struct node ** first)
{
    struct node * temp = NULL;

    if(*first == NULL)
    {
         return ;
    }
    else if( (*first)-> next == NULL)
    {
         free(*first);
         *first = NULL;
    }
    else
    {
        temp = *first;

        while(temp->next->next != NULL)
        {
            temp = temp -> next;
        }
        free (temp->next);
        temp ->next = NULL;

 
    }
}

void display(struct node * first)
{

    while(first != NULL)
    {
        printf("| %d | -> " ,first->data);
        first = first->next;
    }
    printf("NULL\n");
}

int count (struct node * first)
{
    int icount = 0;

    while(first != NULL)
    {
        icount++;
        first = first -> next;
    }
    return icount;
}

void insertatpos(struct node ** first,int no,int pos)
{
    struct node * newn = NULL;
    struct node * temp = NULL;
    int icnt = 0;
    int icount = 0;
  
    icount = count(*first);
    if(pos < 1 || pos > icount+1)
    {
        printf("invalid position \n");
        return;
    }

    if(pos == 1)
    {
        insertfirst(first ,no);
    }
    else if(pos == icount + 1)
    {
        insertlast(first,no);
    }
    else
    {
        temp = *first;

        newn = (struct node *)malloc(sizeof(struct node));
        
        for(icnt = 1; icnt < pos-1;icnt++)
        {
            temp = temp -> next;
        }

        newn-> data = no;
        newn -> next = NULL;

        newn->next = temp->next;
        temp->next = newn;
    }
}

void deleteatpos(struct node ** first ,int pos)
{
    int icount = 0;
    int icnt = 0;
    struct node * temp = NULL;
    struct node * target = NULL;

    icount = count(*first);

    if(pos < 1 || pos > icount )
    {
        printf("invalid position \n");
        return;
    }

    if(pos == 1)
    {
        deletefirst(first);   
    }
    else if(pos == icount)
    {
        deletelast(first);
    }
    else
    {
        temp = *first;

        for(icnt = 0; icnt < pos-1;icnt++)
        {
            temp = temp -> next;
        }
        target = temp -> next;
        temp -> next = target->next;
        free(target);

    }
}



int main()
{
    struct node * head = NULL;
    int iret = 0;

    
    insertfirst(&head,51);
    insertfirst(&head,21);
    insertfirst(&head,11);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    insertlast(&head,101);
    insertlast(&head,111);
    insertlast(&head,121);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    deletefirst(&head);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    deletelast(&head);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    insertatpos(&head,105,5);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    deleteatpos(&head,5);

    display(head);

    iret = count(head);
    printf("element in linkedlist is : %d \n",iret);

    return 0;
}

File: LINKLIST/JAVA/singlyLLL.java
class node
{
    public int data ;
    public node next;
}

class singlyLLLX
{
    private node first;
    private node last ;
    private int icount;

    public
    singlyLLLX()
    {
        first = null;
        last = null;
        icount = 0;
    }

    void insertfirst(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;
    
        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
        icount++;
    }

    void insertlast(int no)
    {
        node newn = null;
        node temp = null;

        newn = new node();

        newn.data = no;
        newn.next = null;
    
        if(first == null)
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null)
            {
                temp = temp.next;
            }
            temp.next = newn;
        }
        icount++;
    }

    void insertatpos(int no,int pos)
    {
        node newn = null;
        node temp = null;
        int icnt = 0;


        icount = count();

        if(pos < 1 || pos > icount + 1)
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            insertfirst(no);
        }
        else if(pos == icount + 1)
        {
            insertlast(no);
        }
        else
        {
            temp = first;
            newn = new node();

            newn.data = no;
            newn.next = null;

            for(icnt = 1;icnt < pos-1 ;icnt++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next= newn;

            icount++;

        }
    }

    void deletefirst()
    {
        if(first == null)
        {
            return;
        }
        else if (first.next == null)
        {
            first = null;
        }
        else
        {
            first = first.next;
        }
        icount--;
    }

    void deletelast()
    {
        node temp = null;

        if(first == null)
        {
            return;
        }
        else if (first.next == null)
        {
            first = null;
        }
        else
        {
            temp = first;

            while(temp.next.next != null)
            {
                temp = temp.next;
            }
            temp.next = null;
        }
        icount--;
    }

    void deleteatpos(int pos)
    {
        node temp = null;
        int icnt = 0;
        icount = count();

        if(pos < 1 || pos > icount )
        {
            System.out.println("invalid position");
            return;
        }

        if(pos == 1)
        {
            deletefirst();
        }
        else if(pos == icount )
        {
            deletelast();
        }
        else
        {
            temp = first;

            for(icnt = 1;icnt < pos - 1;icnt++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;

            icount--;
        }

    }

    void display()
    {
        node temp = null;

        temp = first;

        while(temp != null)
        {
            System.out.print(" | "+temp.data+" | ->");
            temp = temp.next;
        }
        System.out.println("NULL");

    }

    int count ()
    {
        return icount;
    }

}

class singlyLLL
{
    public static void main(String a[])
    {
        singlyLLLX sobj = new singlyLLLX();
        int iret = 0;

        sobj.insertfirst(51);
        sobj.insertfirst(21);
        sobj.insertfirst(11);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertlast(101);
        sobj.insertlast(111);
        sobj.insertlast(121);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.insertatpos(105,5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);


        sobj.deletefirst();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deletelast();

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

        sobj.deleteatpos(5);

        sobj.display();
        iret = sobj.count();
        System.out.println("no of element in linklist : " + iret);

    }
}
